clients/src/main/java/org/apache/kafka/common/utils/Checksums.java

```diff
@@ -40,10 +55,32 @@ public static void update(Checksum checksum, ByteBuffer buffer, int length) {
     public static void update(Checksum checksum, ByteBuffer buffer, int offset, int length) {
         if (buffer.hasArray()) {
             checksum.update(buffer.array(), buffer.position() + buffer.arrayOffset() + offset, length);
-        } else {
-            int start = buffer.position() + offset;
-            for (int i = start; i < start + length; i++)
-                checksum.update(buffer.get(i));
+            return;
+        }
+        if (BYTE_BUFFER_UPDATE != null) {
+            final int oldPosition = buffer.position();
+            final int oldLimit = buffer.limit();
+            try {
+                // save a slice to be used to save an allocation in the hot-path
+                final int start = buffer.position() + offset;
+                buffer.position(0);
+                buffer.limit(start + length);
+                buffer.position(start);
+                BYTE_BUFFER_UPDATE.invokeExact(checksum, buffer);
+                return;
+            } catch (Throwable t) {
+                // fallback
+            } finally {
+                // reset buffer's offsets
+                buffer.position(0);
```

Comment:
Is it intentional that we are calling `position` twice in the `finally` block?

Reply:
I've just read the `ByteBuffer` doc: I have moved first `position` to save `limit`'s change to throw any exception, but I see that it won't happen...let me remove this, good point :+1: